Added shape inference for TFL to Tosa Lowering

rsuderman opened the pull request:
Some TFLite operations need the rank/shape of operands to lower to the
appropriate TOSA operations. By performing shape inference during the
lowering, underspecified IR can be lowered without TFLite shapes known during
compilation.

Note this splits lowering from quantized type removal as the Dialect Conversion
framework does not agree with multiple sources of types.

rsuderman committed:
Added shape inference for TFL to Tosa Lowering

Some TFLite operations need the rank/shape of operands to lower to the
appropriate TOSA operations. By performing shape inference during the
lowering, underspecified IR can be lowered without TFLite shapes known during
compilation.

Note this splits lowering from quantized type removal as the Dialect Conversion
framework does not agree with multiple sources of types.

stellaraccident commented on tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir:
```diff
@@ -1278,3 +1278,15 @@ func @test_fullyconnected_hybrid(%arg0: tensor<14x19xf32>) -> tensor<14x28xf32>
+  %cst = constant dense<0.000000e+00> : tensor<16xf32>
+  // CHECK: tosa.add
+  // CHECK: tosa.conv2d
+  // CHECKL tensor.cast
```
Type: 'CHECKL' -> 'CHECK:'

Can you extend this test? I am guessing that it is supposed to be updating the shape of the: a) conv result, b) return, c) function signature.

Probably fine to verify a&b since the verifier will enforce c.

stellaraccident replied:
Also, looking through the rest of this patch, a lot of ops are touched but only this one test. Wdyt of strengthening the corresponding tests above by changing them to dynamic shaped versions and then making the existing CHECKs more specific to verify the shape. A lot of the tests in this file are a bit weak to my eye, and this could split the difference of improving some of them and adding test coverage for this PR in one go.

rsuderman replied:
Updated tests and lowering where possible to have unknown output shape. There is additional work
required for the rest but this triages down to a select few.

Tests may be failing, working on cleaning up checks.

stellaraccident reviewed (commented):
Quick first pass.

rsuderman committed:
Updated tests for unranked outputs with legalize fixes

Many cases no longer require the output shape to be known at lowering time.
Updated the simple cases to no longer require the output shape and updated
the corresponding tests.

rsuderman committed:
Updated tests to allow tensor.cast to unranked.

Remaining failures are due to TosaOps bug in MLIR core. Assumes
RankedTensorType when computing the quantization information.

jpienaar commented:
It would seem you get all of what CreateAndInfer does if you just used effectively https://github.com/llvm/llvm-project/blob/24c8eaec9467b2aaf70b0db33a4e4dd415139a50/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L136.

rsuderman commented:
> It would seem you get all of what CreateAndInfer does if you just used effectively [llvm/llvm-project@`24c8eae`/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L136](https://github.com/llvm/llvm-project/blob/24c8eaec9467b2aaf70b0db33a4e4dd415139a50/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L136).

So I am not entirely certain that is true. This works for the majority of operations however tosa.rescale depends on the result element type being specified directly by the constructor.

Overall it would be nice to have an interface that supports this, however I don't think the current ones do. E.g. all the existing utilities are member functions as well, requiring a constructed op to reify the shapes.

stellaraccident commented on tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir:
```diff
@@ -12,30 +12,30 @@
 // CHECK-LABEL: test_conv2d
```
Check the func result shape?

stellaraccident replied:
(here and below)

rsuderman replied:
Should be all added in.

stellaraccident commented on tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir:
```diff
@@ -1275,6 +1270,18 @@ func @test_fullyconnected_hybrid(%arg0: tensor<14x19xf32>) -> tensor<14x28xf32>
+
+// -----
+
+func @test_conv2d_infer(%arg0: tensor<1x32x32x8xf32>, %arg1: tensor<16x1x1x8xf32>) -> tensor<*xf32> {
```
Check-label (and verify the func return type)

rsuderman replied:
Check-label added

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -3075,9 +3002,53 @@ void LegalizeTFL::runOnFunction() {
+  // the FuncOp type.
+  IRRewriter rewriter(func.getContext());
+  func.walk([&](ReturnOp op) {
+    FuncOp parent = dyn_cast<FuncOp>(op->getParentOp());
```
auto parent = op->getParentOfType<FuncOp>();

rsuderman replied:
Done.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_tfl.cc:
```diff
@@ -3075,9 +3002,53 @@ void LegalizeTFL::runOnFunction() {
+    bool castAdded = false;
+    SmallVector<Value> castedValues;
+    for (auto it : llvm::zip(op->getOperands(), resultTys)) {
+      auto operand = std::get<0>(it);
```
nit: Can you make these explicitly types (a couple too many levels of auto for my taste)

rsuderman replied:
Auto finished.

stellaraccident commented on tensorflow/compiler/mlir/tosa/transforms/legalize_utils.h:
```diff
@@ -119,6 +121,54 @@ llvm::Optional<Value> getConstTensor(PatternRewriter& rewriter, Operation* op,
+          .failed())
+    return op;
+
+  auto result = op->getResult(0);
```
Can you please provide some justificaiton/example of when inferReturnTypeComponents is not sufficient and must be augmented with the DFA? It isn't clear to me.

rsuderman replied:
Explanation added.

Basically tosa.rescale requires output type information not specified on the operation. Using the existing output type gives us the full type (not possible via pure shape inference). It may make sense to add a TypeAttr on rescales to avoid this foo. It would also allow us to more easily define TOSA models as a list of operations.

jpienaar commented on tensorflow/compiler/mlir/tosa/transforms/legalize_utils.h:
```diff
@@ -119,6 +121,54 @@ llvm::Optional<Value> getConstTensor(PatternRewriter& rewriter, Operation* op,
+
+  // Compute the knowledge based on the inferred type.
+  auto inferredKnowledge = ValueKnowledge::getPessimisticValueState();
+  inferredKnowledge.dtype = result_ty.cast<ShapedType>().getElementType();
```
So this is all to handle rescale? Passing in an unranked tensor with good element type and then utilizing inference to compute the same and then merge? (so the result element type could have been passed in directly and just updated via clone?)

rsuderman replied:
Bingo. Its kinda neat because we can define a model purely as unranked tensors with known element types and *in theory* all shapes will be known by output.

jpienaar commented:
> > It would seem you get all of what CreateAndInfer does if you just used effectively [llvm/llvm-project@`24c8eae`/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L136](https://github.com/llvm/llvm-project/blob/24c8eaec9467b2aaf70b0db33a4e4dd415139a50/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L136).
> 
> So I am not entirely certain that is true. This works for the majority of operations however tosa.rescale depends on the result element type being specified directly by the constructor.

I see, so rescale can change the result element type and the target element type is not an attribute?

> 
> Overall it would be nice to have an interface that supports this, however I don't think the current ones do. E.g. all the existing utilities are member functions as well, requiring a constructed op to reify the shapes.

You are doing construction time inference, so the producers are all created before a new op is. But you are correct that if you want to reify before creating an op, then you can't. But that feels weird to me to do: I see this as straddling line between tensor world and buffer world and should be a dedicated pass to go from one to the other.

rsuderman committed:
Fix some minor issues, add more details on type inference.

rsuderman committed:
Fixed missing check-label

rsuderman committed:
Fixed uint8 pass to support unranked types.

rsuderman commented:
> orts this, however I don't think the current ones do. E.g. all the existing utilities are member functions as well, requiring a constructed op to reify the shapes.

Overall I think the right choice is to affix the output element type as a type attr for tosa.rescale. It actually avoids some other issues as well (u8 needing to persist as a type until tosa-to-linalg).

It may be worth raising with the ARM folks.

rsuderman commented:
Everything should be cleaned up including some changes due to moving the quantization stripping to an optional pass. Verified manually but there may be integration issues needed internally.
